File: app-rating/src/main/java/com/stepstone/apprating/AppRatingDialogFragment.java
/*
Copyright 2017 StepStone Services

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

package com.stepstone.apprating;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;

/**
 * This class represents rating dialog created by {@link com.stepstone.apprating.AppRatingDialog.Builder}.
 *
 * @see AppRatingDialog
 */
public class AppRatingDialogFragment extends DialogFragment {

    public static AppRatingDialogFragment newInstance(AppRatingDialog.Builder.Data data) {
        AppRatingDialogFragment fragment = new AppRatingDialogFragment();
        fragment.setData(data);
        return fragment;
    }

    private AppRatingDialog.Builder.Data data;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return setupAlertDialog(getActivity());
    }

    private AlertDialog setupAlertDialog(Context context) {
        final AppRatingDialogView dialogView = new AppRatingDialogView(context);

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        setupPositiveButton(dialogView, builder);
        setupNegativeButton(builder);
        setupTitleAndContentMessages(dialogView);
        setupColors(dialogView);

        dialogView.setNumberOfStars(data.numberOfStars);

        if (data.noteDescriptions != null && !data.noteDescriptions.isEmpty()) {
            dialogView.setNoteDescriptions(data.noteDescriptions);
        }

        dialogView.setDefaultRating(data.defaultRating);
        builder.setView(dialogView);
        return builder.create();
    }

    private void setupColors(AppRatingDialogView dialogView) {
        if (data.titleColorResId != 0) {
            dialogView.setTitleColor(data.titleColorResId);
        }
        if (data.contentColorResId != 0) {
            dialogView.setContentColor(data.contentColorResId);
        }
    }

    private void setupTitleAndContentMessages(AppRatingDialogView dialogView) {
        if (!TextUtils.isEmpty(getTitle())) {
            dialogView.setTitleText(getTitle());
        }
        if (!TextUtils.isEmpty(getContent())) {
            dialogView.setContentText(getContent());
        }
    }

    private void setupNegativeButton(AlertDialog.Builder builder) {
        if (!TextUtils.isEmpty(getNegativeButtonText())) {
            builder.setNegativeButton(getNegativeButtonText(), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    data.negativeButtonClickedListener.onClicked();
                }
            });
        }
    }

    private void setupPositiveButton(final AppRatingDialogView dialogView, AlertDialog.Builder builder) {
        if (!TextUtils.isEmpty(getPositiveButtonText())) {
            builder.setPositiveButton(getPositiveButtonText(), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    final int rateNumber = (int) dialogView.getRateNumber();
                    final String comment = dialogView.getComment();
                    data.positiveButtonClickedListener.onClicked(rateNumber, comment);
                }
            });
        }
    }

    private String getTitle() {
        if (TextUtils.isEmpty(data.title)) {
            if (data.titleResId == 0) {
                return null;
            }
            return getString(data.titleResId);
        }
        return data.title;
    }

    private String getContent() {
        if (TextUtils.isEmpty(data.content)) {
            if (data.contentResId == 0) {
                return null;
            }
            return getString(data.contentResId);
        }
        return data.content;
    }

    private String getPositiveButtonText() {
        if (TextUtils.isEmpty(data.positiveButtonText)) {
            if (data.positiveButtonTextResId == 0) {
                return null;
            }
            return getString(data.positiveButtonTextResId);
        }
        return data.positiveButtonText;
    }

    private String getNegativeButtonText() {
        if (TextUtils.isEmpty(data.negativeButtonText)) {
            if (data.negativeButtonTextResId == 0) {
                return null;
            }
            return getString(data.negativeButtonTextResId);
        }
        return data.negativeButtonText;
    }

    public void setData(AppRatingDialog.Builder.Data data) {
        this.data = data;
    }
}

File: app-rating/src/main/java/com/stepstone/apprating/AppRatingDialogFragment.kt
/*
Copyright 2017 StepStone Services

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

package com.stepstone.apprating

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.appcompat.app.AlertDialog
import android.text.TextUtils
import com.stepstone.apprating.PreferenceHelper.setAgreeShowDialog
import com.stepstone.apprating.PreferenceHelper.setRemindInterval
import com.stepstone.apprating.extensions.applyIfNotZero
import com.stepstone.apprating.listener.RatingDialogListener

/**
 * This class represents rating dialog created by [com.stepstone.apprating.AppRatingDialog.Builder].

 * @see AppRatingDialog
 */
class AppRatingDialogFragment : DialogFragment() {

    private var listener: RatingDialogListener? = null
        get() {
            if (host is RatingDialogListener) {
                return host as RatingDialogListener
            }
            return targetFragment as RatingDialogListener?
        }

    private lateinit var data: AppRatingDialog.Builder.Data
    private lateinit var alertDialog: AlertDialog
    private lateinit var dialogView: AppRatingDialogView

    private val title by lazy { data.title.resolveText(resources) }
    private val description by lazy { data.description.resolveText(resources) }
    private val defaultComment by lazy { data.defaultComment.resolveText(resources) }
    private val hint by lazy { data.hint.resolveText(resources) }
    private val positiveButtonText by lazy { data.positiveButtonText.resolveText(resources) }
    private val neutralButtonText by lazy { data.neutralButtonText.resolveText(resources) }
    private val negativeButtonText by lazy { data.negativeButtonText.resolveText(resources) }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return setupAlertDialog(activity!!)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putFloat(CURRENT_RATE_NUMBER, dialogView.rateNumber)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val rateNumber: Float? = savedInstanceState?.getFloat(CURRENT_RATE_NUMBER)
        if (rateNumber != null) {
            dialogView.setDefaultRating(rateNumber.toInt())
        }
    }

    private fun setupAlertDialog(context: Context): AlertDialog {
        dialogView = AppRatingDialogView(context)
        val builder = AlertDialog.Builder(activity!!)
        data = arguments?.getSerializable(DIALOG_DATA) as AppRatingDialog.Builder.Data

        setupPositiveButton(dialogView, builder)
        setupNegativeButton(builder)
        setupNeutralButton(builder)
        setupTitleAndContentMessages(dialogView)
        setupHint(dialogView)
        setupColors(dialogView)
        setupInputBox()
        setupRatingBar()

        builder.setView(dialogView)
        alertDialog = builder.create()
        if (data.dialogBackgroundColorResId != 0) {
            alertDialog.window?.setBackgroundDrawableResource(data.dialogBackgroundColorResId)
        }
        setupAnimation()
        setupCancelable()
        return alertDialog
    }

    private fun setupRatingBar() {
        dialogView.setNumberOfStars(data.numberOfStars)

        val isEmpty = data.noteDescriptions?.isEmpty() ?: true
        if (!isEmpty) {
            dialogView.setNoteDescriptions(data.noteDescriptions!!)
        }

        dialogView.setDefaultRating(data.defaultRating)
    }

    private fun setupInputBox() {
        dialogView.setCommentInputEnabled(data.commentInputEnabled)
        dialogView.setThreshold(data.threshold)
    }

    private fun setupCancelable() {
        data.cancelable?.let { isCancelable = it }
        data.canceledOnTouchOutside?.let { alertDialog.setCanceledOnTouchOutside(it) }
    }

    private fun setupAnimation() {
        if (data.windowAnimationResId != 0) {
            alertDialog.window.attributes.windowAnimations = data.windowAnimationResId
        }
    }

    private fun setupColors(dialogView: AppRatingDialogView) {
        data.titleTextColorResId.applyIfNotZero {
            dialogView.setTitleTextColor(this)
        }
        data.descriptionTextColorResId.applyIfNotZero {
            dialogView.setDescriptionTextColor(this)
        }
        data.commentTextColorResId.applyIfNotZero {
            dialogView.setEditTextColor(this)
        }
        data.commentBackgroundColorResId.applyIfNotZero {
            dialogView.setEditBackgroundColor(this)
        }
        data.hintTextColorResId.applyIfNotZero {
            dialogView.setHintColor(this)
        }
        data.starColorResId.applyIfNotZero {
            dialogView.setStarColor(this)
        }
        data.noteDescriptionTextColor.applyIfNotZero {
            dialogView.setNoteDescriptionTextColor(this)
        }
    }

    private fun setupTitleAndContentMessages(dialogView: AppRatingDialogView) {
        if (!title.isNullOrEmpty()) {
            dialogView.setTitleText(title!!)
        }
        if (!description.isNullOrEmpty()) {
            dialogView.setDescriptionText(description!!)
        }
        if (!defaultComment.isNullOrEmpty()) {
            dialogView.setDefaultComment(defaultComment!!)
        }
    }

    private fun setupHint(dialogView: AppRatingDialogView) {
        if (!TextUtils.isEmpty(hint)) {
            dialogView.setHint(hint!!)
        }
    }

    private fun setupNegativeButton(builder: AlertDialog.Builder) {
        if (!TextUtils.isEmpty(negativeButtonText)) {
            builder.setNegativeButton(negativeButtonText) { _, _ ->
                setAgreeShowDialog(context, false)
                listener?.onNegativeButtonClicked()
            }
        }
    }

    private fun setupPositiveButton(dialogView: AppRatingDialogView, builder: AlertDialog.Builder) {
        if (!TextUtils.isEmpty(positiveButtonText)) {
            builder.setPositiveButton(positiveButtonText) { _, _ ->
                val rateNumber = dialogView.rateNumber.toInt()
                setAgreeShowDialog(context, false)
                if (rateNumber <= data.threshold) {
                    val comment = dialogView.comment
                    listener?.onPositiveButtonClickedWithComment(rateNumber, comment)
                } else { //rating>threshold
                    listener?.onPositiveButtonClickedWithoutComment(rateNumber)
                }
            }
        }
    }

    private fun setupNeutralButton(builder: AlertDialog.Builder) {
        if (!TextUtils.isEmpty(neutralButtonText)) {
            builder.setNeutralButton(neutralButtonText) { _, _ ->
                setRemindInterval(context)
                listener?.onNeutralButtonClicked()
            }
        }
    }

    companion object {

        fun newInstance(data: AppRatingDialog.Builder.Data): AppRatingDialogFragment {
            val fragment = AppRatingDialogFragment()
            val bundle = Bundle()
            bundle.putSerializable(DIALOG_DATA, data)
            fragment.arguments = bundle
            return fragment
        }
    }
}
